crates/isograph_compiler/src/watch.rs

```diff
@@ -52,20 +56,102 @@ pub async fn handle_watch_command(
     .await
 }
 
-fn any_modified_path_is_outside_artifact_directory(
-    events: &[DebouncedEvent],
-    artifact_directory: &PathBuf,
-) -> bool {
-    for event in events.iter() {
-        for watched_path in event.paths.iter() {
-            // These paths are canonicalized, so it is sufficient
-            // to check starts_with
-            if !watched_path.starts_with(artifact_directory) {
-                return true;
+fn compile_and_print(state: &mut CompilerState, changes: &[SourceFileEvent]) {
+    eprintln!("{}", "Starting to compile.".cyan());
+    let result = WithDuration::new(|| state.compile(changes));
+    let _ = print_result(result.item, result.elapsed_time);
+}
+
+fn changed_files(events: &[DebouncedEvent], config: &CompilerConfig) -> Vec<SourceFileEvent> {
+    events
+        .iter()
+        .filter_map(|event| match event.kind {
+            EventKind::Create(create_kind) => {
+                process_create_event(config, create_kind, &event.paths)
+            }
+            EventKind::Modify(modify_kind) => {
+                process_modify_event(config, modify_kind, &event.paths)
+            }
+            EventKind::Remove(remove_kind) => {
+                process_remove_event(config, remove_kind, &event.paths)
             }
+            _ => None,
+        })
+        .collect::<Vec<_>>()
+}
+
+fn process_create_event(
+    config: &CompilerConfig,
+    create_kind: CreateKind,
+    paths: &[PathBuf],
+) -> Option<SourceFileEvent> {
+    match create_kind {
+        CreateKind::File if paths.len() == 1 => changed_file_kind(config, &paths[0])
```

Comment:
what happens if you change branches? presumably there are a bajillion create events at the same time... might be worth noting that as a TODO, or perhaps having some sort of logic where any event we don't handle just results in `read_and_parse_all` getting called.

Reply:
Yeah, fallback is a good idea

Reply:
This sounds good to me.

On a side note, watchman integrates with source control. But it doesn't look like that's exposed via `notify-debouncer`. Anyway, this sounds good for now and we can always revisit.